File: pipelines/lib/arcgis.ts
/**
 * Minimal ArcGIS Feature Service client, built for verifiable bulk extraction.
 *
 * Extraction strategy, in the order the prompt requires:
 *   1. read service and layer metadata
 *   2. ask for the full OBJECTID list (returnIdsOnly=true)
 *   3. download features in bounded OBJECTID batches, respecting maxRecordCount
 *   4. retry with exponential backoff
 *   5. reconcile downloaded ids against the requested id list
 *
 * Batching by explicit id ranges rather than resultOffset paging is deliberate.
 * Offset paging over a service that is being edited underneath you can silently
 * skip or duplicate rows; an id list pins exactly which features are expected,
 * so a shortfall is detectable rather than invisible.
 */

import { createHash } from 'node:crypto';

export interface ArcGisError extends Error {
  httpStatus?: number;
  esriCode?: number;
}

export interface FetchJsonOptions {
  retries?: number;
  timeoutMs?: number;
  label?: string;
  /**
   * Form-encoded body. When present the request is POSTed. Batched OBJECTID
   * lists routinely exceed the ~8 KB URL limit and a GET returns HTTP 414, so
   * every query that carries an id list must POST.
   */
  body?: URLSearchParams;
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

export async function fetchJson<T = any>(
  url: string,
  opts: FetchJsonOptions = {},
): Promise<T> {
  const retries = opts.retries ?? 5;
  const timeoutMs = opts.timeoutMs ?? 120_000;
  let lastErr: unknown;

  for (let attempt = 0; attempt <= retries; attempt++) {
    if (attempt > 0) {
      const backoff = Math.min(30_000, 1000 * 2 ** (attempt - 1));
      await sleep(backoff + Math.floor(Math.random() * 250));
    }
    const ac = new AbortController();
    const timer = setTimeout(() => ac.abort(), timeoutMs);
    try {
      const res = await fetch(url, {
        signal: ac.signal,
        ...(opts.body
          ? {
              method: 'POST',
              body: opts.body,
              headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            }
          : {}),
      });
      if (!res.ok) {
        const e = new Error(
          `HTTP ${res.status} for ${opts.label ?? url}`,
        ) as ArcGisError;
        e.httpStatus = res.status;
        throw e;
      }
      const body = (await res.json()) as any;
      // ArcGIS returns HTTP 200 with an error envelope. Never treat that as data.
      if (body && body.error) {
        const e = new Error(
          `ArcGIS error ${body.error.code}: ${body.error.message}` +
            (body.error.details?.length ? ` (${body.error.details.join('; ')})` : ''),
        ) as ArcGisError;
        e.esriCode = body.error.code;
        throw e;
      }
      return body as T;
    } catch (err) {
      lastErr = err;
    } finally {
      clearTimeout(timer);
    }
  }
  throw lastErr instanceof Error ? lastErr : new Error(String(lastErr));
}

export function queryUrl(
  serviceUrl: string,
  layerId: number,
  params: Record<string, string | number | boolean | undefined>,
): string {
  const u = new URL(`${serviceUrl}/${layerId}/query`);
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined) u.searchParams.set(k, String(v));
  }
  return u.toString();
}

/** POST a query. Use this whenever parameters may be large (id lists, geometry). */
export function queryBody(
  params: Record<string, string | number | boolean | undefined>,
): URLSearchParams {
  const b = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined) b.set(k, String(v));
  }
  return b;
}

export interface LayerMeta {
  id: number;
  name: string;
  type: string;
  geometryType?: string;
  objectIdField: string;
  globalIdField?: string;
  maxRecordCount: number;
  supportsPagination: boolean;
  supportsStatistics: boolean;
  fields: { name: string; type: string; alias?: string; domain?: any }[];
  extent?: any;
  copyrightText?: string;
  description?: string;
  raw: any;
}

export async function getLayerMeta(
  serviceUrl: string,
  layerId: number,
): Promise<LayerMeta> {
  const raw = await fetchJson(`${serviceUrl}/${layerId}?f=json`, {
    label: `layer ${layerId} metadata`,
  });
  return {
    id: raw.id,
    name: raw.name,
    type: raw.type,
    geometryType: raw.geometryType,
    objectIdField: raw.objectIdField ?? 'OBJECTID',
    globalIdField: raw.globalIdField,
    maxRecordCount: raw.maxRecordCount ?? 1000,
    supportsPagination: Boolean(raw.advancedQueryCapabilities?.supportsPagination),
    supportsStatistics: Boolean(raw.advancedQueryCapabilities?.supportsStatistics),
    fields: raw.fields ?? [],
    extent: raw.extent,
    copyrightText: raw.copyrightText,
    description: raw.description,
    raw,
  };
}

export interface ExtentFilter {
  xmin: number;
  ymin: number;
  xmax: number;
  ymax: number;
  wkid: number;
}

export interface CountOptions {
  where: string;
  extent?: ExtentFilter | null;
}

export async function getCount(
  serviceUrl: string,
  layerId: number,
  o: CountOptions,
): Promise<number> {
  const res = await fetchJson<{ count: number }>(
    `${serviceUrl}/${layerId}/query`,
    {
      label: `layer ${layerId} count`,
      body: queryBody({
        where: o.where,
        returnCountOnly: true,
        f: 'json',
        ...extentParams(o.extent),
      }),
    },
  );
  return res.count;
}

export async function getObjectIds(
  serviceUrl: string,
  layerId: number,
  o: CountOptions,
): Promise<number[]> {
  const res = await fetchJson<{ objectIds: number[] | null }>(
    `${serviceUrl}/${layerId}/query`,
    {
      label: `layer ${layerId} object ids`,
      body: queryBody({
        where: o.where,
        returnIdsOnly: true,
        f: 'json',
        ...extentParams(o.extent),
      }),
    },
  );
  return res.objectIds ?? [];
}

function extentParams(
  e: ExtentFilter | null | undefined,
): Record<string, string | undefined> {
  if (!e) return {};
  return {
    geometry: JSON.stringify({
      xmin: e.xmin,
      ymin: e.ymin,
      xmax: e.xmax,
      ymax: e.ymax,
      spatialReference: { wkid: e.wkid },
    }),
    geometryType: 'esriGeometryEnvelope',
    inSR: String(e.wkid),
    spatialRel: 'esriSpatialRelIntersects',
  };
}

export interface DownloadOptions {
  serviceUrl: string;
  layerId: number;
  objectIds: number[];
  outFields: string;
  returnGeometry: boolean;
  outSR: number;
  batchSize: number;
  concurrency?: number;
  objectIdField?: string;
  onProgress?: (done: number, total: number) => void;
}

export interface DownloadResult {
  features: any[];
  /** SHA-256 over every page payload, in batch order. Reproducible. */
  sha256: string;
  batches: number;
  /** Ids that were requested but never came back. Empty on a clean extract. */
  missingIds: number[];
  duplicateIds: number[];
}

/**
 * Downloads features by explicit OBJECTID batches and reconciles the result.
 * Throws only on unrecoverable transport failure; a short extract is reported
 * through `missingIds` so the caller can mark the snapshot partial rather than
 * silently shipping a hole in the network.
 */
export async function downloadByIds(o: DownloadOptions): Promise<DownloadResult> {
  const idField = o.objectIdField ?? 'OBJECTID';
  const batchSize = Math.max(1, o.batchSize);
  const concurrency = Math.max(1, o.concurrency ?? 4);

  // Reconciliation reads the OBJECTID off each returned feature, so the id
  // field has to be requested even when the caller only wants a few columns.
  // Without this every id looks "missing" and a clean extract reports as short.
  const outFields =
    o.outFields === '*' ||
    o.outFields.split(',').some((f) => f.trim().toLowerCase() === idField.toLowerCase())
      ? o.outFields
      : `${idField},${o.outFields}`;

  const batches: number[][] = [];
  for (let i = 0; i < o.objectIds.length; i += batchSize) {
    batches.push(o.objectIds.slice(i, i + batchSize));
  }

  const pages: (any[] | undefined)[] = new Array(batches.length);
  const hashes: (string | undefined)[] = new Array(batches.length);
  let done = 0;
  let next = 0;

  const worker = async () => {
    for (;;) {
      const idx = next++;
      if (idx >= batches.length) return;
      // POST: an OBJECTID batch of 2000 ids far exceeds the GET URL limit.
      const body = await fetchJson<any>(`${o.serviceUrl}/${o.layerId}/query`, {
        label: `batch ${idx + 1}/${batches.length}`,
        body: queryBody({
          objectIds: batches[idx].join(','),
          outFields,
          returnGeometry: o.returnGeometry,
          outSR: o.outSR,
          f: 'json',
        }),
      });
      const feats = body.features ?? [];
      pages[idx] = feats;
      hashes[idx] = createHash('sha256').update(JSON.stringify(body)).digest('hex');
      done += feats.length;
      o.onProgress?.(done, o.objectIds.length);
    }
  };

  await Promise.all(Array.from({ length: concurrency }, worker));

  const features: any[] = [];
  for (const p of pages) if (p) features.push(...p);

  // --- reconcile ---
  const seen = new Map<number, number>();
  for (const f of features) {
    const oid = f.attributes?.[idField];
    seen.set(oid, (seen.get(oid) ?? 0) + 1);
  }
  const missingIds = o.objectIds.filter((id) => !seen.has(id));
  const duplicateIds = [...seen.entries()].filter(([, c]) => c > 1).map(([id]) => id);

  const roll = createHash('sha256');
  for (const h of hashes) roll.update(h ?? '');

  return {
    features,
    sha256: roll.digest('hex'),
    batches: batches.length,
    missingIds,
    duplicateIds,
  };
}

File: python/src/nzcl/arcgis.py
"""Minimal ArcGIS Feature Service client, built for verifiable bulk extraction.

Strategy:
  1. read service and layer metadata
  2. ask for the full OBJECTID list (returnIdsOnly=true)
  3. download in bounded OBJECTID batches, respecting maxRecordCount
  4. retry with exponential backoff
  5. reconcile what came back against what was asked for

Batching by explicit id list rather than resultOffset paging is deliberate.
Offset paging over a service being edited underneath can silently skip or
duplicate rows; an id list pins exactly which features are expected, so a
shortfall is detectable rather than invisible.

Every query is POSTed. A batch of 2000 OBJECTIDs in a query string exceeds the
URL length limit and the service returns HTTP 414.
"""

from __future__ import annotations

import hashlib
import json
import random
import time
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass, field
from typing import Any, Callable, Iterable, Sequence

import requests


class ArcGisError(RuntimeError):
    def __init__(self, message: str, *, http_status: int | None = None,
                 esri_code: int | None = None) -> None:
        super().__init__(message)
        self.http_status = http_status
        self.esri_code = esri_code


_session = requests.Session()


def fetch_json(
    url: str,
    *,
    data: dict[str, Any] | None = None,
    retries: int = 5,
    timeout: float = 120.0,
    label: str | None = None,
) -> dict[str, Any]:
    """GET or POST (when `data` is given), with retry and Esri error handling."""
    last: Exception | None = None
    for attempt in range(retries + 1):
        if attempt:
            time.sleep(min(30.0, 1.0 * 2 ** (attempt - 1)) + random.random() * 0.25)
        try:
            if data is None:
                resp = _session.get(url, timeout=timeout)
            else:
                resp = _session.post(url, data=data, timeout=timeout)
            if resp.status_code != 200:
                raise ArcGisError(
                    f"HTTP {resp.status_code} for {label or url}",
                    http_status=resp.status_code,
                )
            body = resp.json()
            # ArcGIS returns HTTP 200 with an error envelope. Never treat that
            # as data.
            if isinstance(body, dict) and "error" in body:
                err = body["error"]
                details = "; ".join(err.get("details") or [])
                raise ArcGisError(
                    f"ArcGIS error {err.get('code')}: {err.get('message')}"
                    + (f" ({details})" if details else ""),
                    esri_code=err.get("code"),
                )
            return body
        except Exception as exc:  # noqa: BLE001 - retried, then re-raised
            last = exc
    raise last if last else ArcGisError("unreachable")


@dataclass
class LayerMeta:
    id: int
    name: str
    type: str
    geometry_type: str | None
    object_id_field: str
    max_record_count: int
    supports_pagination: bool
    fields: list[dict[str, Any]]
    raw: dict[str, Any] = field(repr=False, default_factory=dict)


def get_layer_meta(service_url: str, layer_id: int) -> LayerMeta:
    raw = fetch_json(f"{service_url}/{layer_id}?f=json", label=f"layer {layer_id} meta")
    return LayerMeta(
        id=raw.get("id", layer_id),
        name=raw.get("name", ""),
        type=raw.get("type", ""),
        geometry_type=raw.get("geometryType"),
        object_id_field=raw.get("objectIdField") or "OBJECTID",
        max_record_count=raw.get("maxRecordCount") or 1000,
        supports_pagination=bool(
            (raw.get("advancedQueryCapabilities") or {}).get("supportsPagination")
        ),
        fields=raw.get("fields") or [],
        raw=raw,
    )


def _extent_params(extent: dict[str, float] | None, wkid: int = 2193) -> dict[str, str]:
    if not extent:
        return {}
    return {
        "geometry": json.dumps(
            {
                "xmin": extent["xmin"], "ymin": extent["ymin"],
                "xmax": extent["xmax"], "ymax": extent["ymax"],
                "spatialReference": {"wkid": wkid},
            }
        ),
        "geometryType": "esriGeometryEnvelope",
        "inSR": str(wkid),
        "spatialRel": "esriSpatialRelIntersects",
    }


def get_count(service_url: str, layer_id: int, where: str,
              extent: dict[str, float] | None = None) -> int:
    body = fetch_json(
        f"{service_url}/{layer_id}/query",
        data={"where": where, "returnCountOnly": "true", "f": "json",
              **_extent_params(extent)},
        label=f"layer {layer_id} count",
    )
    return int(body.get("count", 0))


def get_object_ids(service_url: str, layer_id: int, where: str,
                   extent: dict[str, float] | None = None) -> list[int]:
    body = fetch_json(
        f"{service_url}/{layer_id}/query",
        data={"where": where, "returnIdsOnly": "true", "f": "json",
              **_extent_params(extent)},
        label=f"layer {layer_id} object ids",
    )
    return list(body.get("objectIds") or [])


@dataclass
class DownloadResult:
    features: list[dict[str, Any]]
    sha256: str
    batches: int
    missing_ids: list[int]
    duplicate_ids: list[int]


def download_by_ids(
    service_url: str,
    layer_id: int,
    object_ids: Sequence[int],
    *,
    out_fields: Iterable[str],
    return_geometry: bool,
    out_sr: int,
    batch_size: int,
    object_id_field: str = "OBJECTID",
    concurrency: int = 6,
    on_progress: Callable[[int, int], None] | None = None,
) -> DownloadResult:
    """Download features by explicit OBJECTID batches, then reconcile.

    A shortfall is reported through `missing_ids` so the caller can mark the
    snapshot partial rather than silently shipping a hole in the network.
    """
    fields = list(out_fields)
    # Reconciliation reads the OBJECTID off each returned feature, so the id
    # field must be requested even when the caller wants only a few columns.
    if object_id_field not in fields and "*" not in fields:
        fields.insert(0, object_id_field)

    ids = list(object_ids)
    batches = [ids[i:i + batch_size] for i in range(0, len(ids), max(1, batch_size))]
    pages: list[list[dict[str, Any]] | None] = [None] * len(batches)
    hashes: list[str] = [""] * len(batches)
    done = 0

    def work(idx: int) -> None:
        nonlocal done
        body = fetch_json(
            f"{service_url}/{layer_id}/query",
            data={
                "objectIds": ",".join(str(i) for i in batches[idx]),
                "outFields": ",".join(fields),
                "returnGeometry": "true" if return_geometry else "false",
                "outSR": str(out_sr),
                "f": "json",
            },
            label=f"batch {idx + 1}/{len(batches)}",
        )
        feats = body.get("features") or []
        pages[idx] = feats
        hashes[idx] = hashlib.sha256(
            json.dumps(body, sort_keys=True).encode()
        ).hexdigest()
        done += len(feats)
        if on_progress:
            on_progress(done, len(ids))

    if batches:
        with ThreadPoolExecutor(max_workers=max(1, concurrency)) as pool:
            list(pool.map(work, range(len(batches))))

    features: list[dict[str, Any]] = []
    for page in pages:
        if page:
            features.extend(page)

    seen: dict[int, int] = {}
    for f in features:
        oid = (f.get("attributes") or {}).get(object_id_field)
        seen[oid] = seen.get(oid, 0) + 1
    missing = [i for i in ids if i not in seen]
    duplicates = [i for i, c in seen.items() if c > 1]

    roll = hashlib.sha256()
    for h in hashes:
        roll.update(h.encode())

    return DownloadResult(
        features=features,
        sha256=roll.hexdigest(),
        batches=len(batches),
        missing_ids=missing,
        duplicate_ids=duplicates,
    )
